### PR TITLE
pip_api: type hints

### LINKED ISSUE #96: Type hints?
`pip-audit` is consuming `pip-api`, and is aiming for 100% MyPy coverage in the process.

`pip-api` currently doesn't have any type signatures, so we disable our MyPy checking wherever we `import pip_api`.

Is adding type signatures to the public APIs worth it here? If so, it's something we could dedicate engineering time to.

### PR DESCRIPTION
**Pull request by woodruffw**
I believe this covers the majority of the public API surface; will test these hints against my work in `pip-audit`.

Closes #96.

### CODE REVIEW

**Comment by woodruffw** [comment]
Turns out that this is a significant pain due to the vendored modules:

```
$ mypy pip_api
pip_api/_vendor/pyparsing.py:123: error: Module "itertools" has no attribute "ifilterfalse"; maybe "filterfalse"?
pip_api/_vendor/pyparsing.py:123: error: Name "filterfalse" already defined (possibly by an import)
pip_api/_vendor/pyparsing.py:126: error: Module "_thread" has no attribute "RLock"
pip_api/_vendor/pyparsing.py:143: error: Cannot find implementation or library stub for module named "ordereddict"
pip_api/_vendor/pyparsing.py:143: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
pip_api/_vendor/pyparsing.py:143: error: Name "_OrderedDict" already defined (possibly by an import)
pip_api/_vendor/pyparsing.py:145: error: Cannot assign to a type
pip_api/_vendor/pyparsing.py:145: error: Incompatible types in assignment (expression has type "None", variable has type "Type[OrderedDict[Any, Any]]")
pip_api/_vendor/pyparsing.py:150: error: Name "SimpleNamespace" already defined (possibly by an import)
pip_api/_vendor/pyparsing.py:233: error: Module has no attribute "maxint"
pip_api/_vendor/pyparsing.py:234: error: Name "xrange" is not defined
pip_api/_vendor/pyparsing.py:236: error: Name "_ustr" already defined on line 227
pip_api/_vendor/pyparsing.py:260: error: Cannot find implementation or library stub for module named "__builtin__"
pip_api/_vendor/pyparsing.py:1800: error: Name "_FifoCache" already defined on line 1771
pip_api/_vendor/pyparsing.py:1829: error: Need type annotation for "packrat_cache" (hint: "packrat_cache: Dict[<type>, <type>] = ...")
pip_api/_vendor/pyparsing.py:2902: error: "Type[ParserElement]" has no attribute "_literalStringClass"
pip_api/_vendor/pyparsing.py:5086: error: Not all arguments converted during string formatting
pip_api/_vendor/pyparsing.py:5944: error: "Callable[[VarArg(Any), KwArg(Any)], Any]" has no attribute "ANY_VALUE"
pip_api/_vendor/pyparsing.py:6730: error: Need type annotation for "_ranges" (hint: "_ranges: List[<type>] = ...")
pip_api/_vendor/pyparsing.py:6799: error: Need type annotation for "_ranges" (hint: "_ranges: List[<type>] = ...")
Found 19 errors in 1 file (checked 21 source files)
```

The CI will pass above, but the parent call to `mypy` in `pip-audit` will continue to fail because of these inner errors. I've tried to mark the imports for `pyparsing` with `# type: ignore`, but that doesn't seem to have done anything...

**Comment by woodruffw** [comment]
Urgh, this might actually be fixable with `py.typed`. Will dig in further tomorrow.

**Comment by woodruffw** [comment]
Okay: in local testing, adding `py.typed` to the distribution allows `pip-audit` to typecheck against its usage of `pip-api`, since said usage is limited to the public APIs that we've annotated.

For the package itself, `mypy` in the CI still won't work, for the vendoring reasons mentioned above. If that's acceptable, this should be good to go 🙂

File: setup.py
Comment (woodruffw) [inline setup.py:31]: N.B.: I'm not 100% sure if this is necessary, but it's what `mypy`'s docs recommend when including `py.typed`.

Comment (di) [reply setup.py:31]: The `zip_safe` thing shouldn't be necessary here: https://github.com/python/mypy/issues/8802

**Comment by di** [comment]
> For the package itself, mypy in the CI still won't work, for the vendoring reasons mentioned above. If that's acceptable, this should be good to go 🙂

I think that's acceptable for now.

**Comment by woodruffw** [comment]
Okay, this should be good to go once the CI goes through!

A minor bump once it's merged would also be appreciated 🙂

**Review by di — APPROVED** [review]

**Comment by di** [comment]
Released in https://pypi.org/project/pip-api/0.0.21/

**Pull request merged**
